File: lib/diamond_square.rb
class DiamondSquare
  def self.rando
    rand() - 0.5
  end
  
  def self.go(times)
    arrays = [[0.5]]
    
    ratio = 2
    
    times.times do
      arrays.map! do |array|
        insert_nils(array)
      end
      arrays = insert_arrays(arrays)
      compute_from_diagonals(arrays) {|a, b, c, d| (a + b + c + d)/4 + rando*ratio} #puts "%1.3f, %1.3f, %1.3f, %1.3f" % [a,b,c,d]; 
      compute_from_adjacents(arrays) {|a, b, c, d| (a + b + c + d)/4 + rando*ratio}
      ratio *= 0.5
    end
    
    return arrays
  end
  
  def self.insert_arrays(arrays)
    new_arrays = []
    arrays.size.times do |i|
      array = arrays[i]
      new_arrays.push array, Array.new(array.size, 0.0)
    end
    return new_arrays
  end
  
  def self.insert_nils(array)
    new_array = []
    array.size.times do |i|
      new_array.push(array[i], 0.0)
    end
    return new_array
  end
  
  def self.compute_from_adjacents(arrays)
    n = arrays.size
    n.times do |row|
      n.times do |col|
        next if (row + col) % 2 == 0
        arrays[row][col] = 
          yield(arrays[(row-1)%n][col], arrays[row][(col-1)%n],
                arrays[row][(col+1)%n], arrays[(row+1)%n][col])
      end
    end
  end
  
  def self.compute_from_diagonals(arrays)
    n = arrays.size
    n.times do |row|
      next if row % 2 == 0
      n.times do |col|
        next if col % 2 == 0
        arrays[row][col] = 
          yield(arrays[(row-1)%n][(col-1)%n], arrays[(row-1)%n][(col+1)%n],
                arrays[(row+1)%n][(col-1)%n], arrays[(row+1)%n][(col+1)%n])
      end
    end
  end
  
  def self.print_arrays(arrays)
    i = 0
    arrays.each do |array|
      print "%4d: " % i
      array.each do |ele|
        print '%1.3f ' % ele
      end
      i += 1
      puts
    end
    puts '---------------------------------------'
  end
end

if __FILE__ == $0
  require 'profiler'
  Profiler__::start_profile
  DiamondSquare.print_arrays(DiamondSquare.go(5))
  Profiler__::print_profile($>)
end

File: boomhammer/public/javascripts/fortress/diamond_square.js
/*global rand */

/*global DiamondSquare */
var DiamondSquare = function(times) {
  function perturbation() {
    return rand() - 0.5;
  }

  function insertNils(array) {
    var newArray = [];
    
    array.length.times(function(i) {
      newArray.push(array[i], 0.0);
    });

    return newArray;
  }

  function insertArrays(arrays) {
    var newArrays = [];

    arrays.length.times(function(i) {
      var a = arrays[i].map(function() { return 0; });
      newArrays.push(arrays[i], a);
    });
    
    return newArrays;
  }

  function computeFromAdjacents(arrays, iterator) {
    var n = arrays.length;
    n.times(function(row) {
      n.times(function(col) {
        if((row + col) % 2 === 0) {
          return; // NEXT
        }

        arrays[row][col] =
          iterator.call(null,
            arrays[Math.mod((row-1), n)][col],
            arrays[row][Math.mod((col-1), n)],
            arrays[row][Math.mod((col+1), n)],
            arrays[Math.mod((row+1), n)][col]
          );
      });
    });
  }

  function computeFromDiagonals(arrays, iterator) {
    var n = arrays.length;
    n.times(function(row) {
      if(row % 2 === 0) {
        return; // NEXT
      }

      n.times(function(col) {
        if(col % 2 === 0) {
          return; // NEXT
        }

        arrays[row][col] =
          iterator.call(null,
            arrays[Math.mod((row-1), n)][Math.mod((col-1), n)],
            arrays[Math.mod((row-1), n)][Math.mod((col+1), n)],
            arrays[Math.mod((row+1), n)][Math.mod((col-1), n)],
            arrays[Math.mod((row+1), n)][Math.mod((col+1), n)]
          );
      });
    });
  }

  function create(times) {
    var arrays = [[0.5]];

    var ratio = 2;

    times.times(function() {
      arrays = arrays.map(function(array) {
        return insertNils(array);
      });

      arrays = insertArrays(arrays);

      computeFromDiagonals(arrays, function(a, b, c, d) {
        return (a + b + c + d)/4 + perturbation()*ratio;
      });
      computeFromAdjacents(arrays, function(a, b, c, d) {
        return (a + b + c + d)/4 + perturbation()*ratio;
      });

      ratio *= 0.5;
    });

    return arrays;
  }

  return create(times);
};